aptos-move/aptos-gas/src/algebra.rs

```diff
@@ -0,0 +1,226 @@
+#[derive(Serialize, Deserialize)]
+pub struct GasQuantity<U> {
+    val: u64,
+    phantom: PhantomData<U>,
+}
+
+pub type Gas = GasQuantity<GasUnit>;
+
+pub type InternalGas = GasQuantity<InternalGasUnit>;
+
+pub type NumBytes = GasQuantity<Byte>;
+
+pub type InternalGasPerByte = GasQuantity<UnitDiv<InternalGasUnit, Byte>>;
+
+pub type InternalGasPerAbstractMemoryUnit =
+    GasQuantity<UnitDiv<InternalGasUnit, AbstractMemoryUnit>>;
+
+pub type AbstractMemorySize = GasQuantity<AbstractMemoryUnit>;
+
+pub type GasScalingFactor = GasQuantity<UnitDiv<InternalGasUnit, GasUnit>>;
+
+pub type Fee = GasQuantity<Octa>;
+
+pub type FeePerGasUnit = GasQuantity<UnitDiv<Octa, GasUnit>>;
+
+/***************************************************************************************************
+ * Constructors
+ *
+ **************************************************************************************************/
+impl<U> GasQuantity<U> {
+    pub fn new(val: u64) -> Self {
+        Self {
+            val,
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn zero() -> Self {
+        Self::new(0)
+    }
+}
+
+/***************************************************************************************************
+ * Conversion
+ *
+ **************************************************************************************************/
+impl<U> From<u64> for GasQuantity<U> {
+    fn from(val: u64) -> Self {
+        Self::new(val)
+    }
+}
+
+// TODO(Gas): This allows the gas value to escape the monad, which weakens the type-level
+//            protection it provides. It's currently needed for practical reasons but
+//            we should look for ways to get rid of it.
+impl<U> From<GasQuantity<U>> for u64 {
+    fn from(gas: GasQuantity<U>) -> Self {
+        gas.val
+    }
+}
+
+/***************************************************************************************************
+ * Clone & Copy
+ *
+ **************************************************************************************************/
+impl<U> Clone for GasQuantity<U> {
+    fn clone(&self) -> Self {
+        Self::new(self.val)
+    }
+}
+
+impl<U> Copy for GasQuantity<U> {}
+
+/***************************************************************************************************
+ * Display & Debug
+ *
+ **************************************************************************************************/
+impl<U> Display for GasQuantity<U> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.val)
+    }
+}
+
+impl<U> Debug for GasQuantity<U> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{} ({})", self.val, std::any::type_name::<U>())
+    }
+}
+
+/***************************************************************************************************
+ * Comparison
+ *
+ **************************************************************************************************/
+impl<U> GasQuantity<U> {
```

Comment:
Why can't we derive all this? You can make a dummy derive on the unit phantom types.